AWS SQS: Source - Allow configuring visibility timeout

barakbaronk opened the pull request:
Hi, first time contributing here, so hopefully I didn't miss anything from the instructions.
In addition, I signed the CLA

**Feature**: 
Allow setting the visibility timeout parameter in the SQS `Receive` request.
this means that in each `Receive` request it will override the default visibility timeout from SQS queue configuration. if one will not set it, the default will be taken.

The requirement:
if one knows the maximum time it takes to process a message and want to set it in the application (and not in the SQS queue console) this change will reduce the number of calls. So, instead of 2 calls - `Receive` and `ChangeVisibilityTimeout`, we will just run the `Receive`

We, At the company I'm working for, started a code migration from our old SQS client to Alpakka.
during the migration we found out this issue, which required us to run 2 consecutive calls - receive the message and change the visibility timeout. I think it's an overhead we can reduce.

I added tests dedicate for this feature
Finally, I looked at the issue tracker and PRs and didn't see any comment about this feature.

ennru commented on sqs/src/main/scala/akka/stream/alpakka/sqs/SqsSourceSettings.scala:
```diff
@@ -16,7 +16,8 @@ final class SqsSourceSettings private (
     val messageAttributeNames: immutable.Seq[MessageAttributeName],
-    val closeOnEmptyReceive: Boolean
+    val closeOnEmptyReceive: Boolean,
+    val visibilityTimeout: Option[Int]
```
I'd rather see this as a `FiniteDuration`. Possibly representing `None` with `Duration.Zero` instead.

barakbaronk replied:
it might be a problem setting the default to `Zero`, since it's a valid value (see [doc](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html#terminating-message-visibility-timeout))
so if one does not want to use this feature, we won't override the default value in SQS queue console.

ennru commented on sqs/src/main/scala/akka/stream/alpakka/sqs/SqsSourceSettings.scala:
```diff
@@ -113,6 +125,7 @@ final class SqsSourceSettings private (
     s"attributeNames=${attributeNames.mkString(",")}, " +
     s"messageAttributeNames=${messageAttributeNames.mkString(",")}, " +
     s"closeOnEmptyReceive=$closeOnEmptyReceive" +
```
```suggestion
     s"closeOnEmptyReceive=$closeOnEmptyReceive," +
```

ennru commented on sqs/src/main/scala/akka/stream/alpakka/sqs/SqsSourceSettings.scala:
```diff
@@ -113,6 +125,7 @@ final class SqsSourceSettings private (
     s"attributeNames=${attributeNames.mkString(",")}, " +
     s"messageAttributeNames=${messageAttributeNames.mkString(",")}, " +
     s"closeOnEmptyReceive=$closeOnEmptyReceive" +
+    s"visibilityTomeout=$visibilityTimeout" +
```
```suggestion
    s"visibilityTimeout=${visibilityTimeout.toCoarsest}" +
```

ennru commented on sqs/src/main/scala/akka/stream/alpakka/sqs/impl/SqsSourceStage.scala:
```diff
@@ -116,3 +119,11 @@ import scala.collection.JavaConverters._
+
+object SqsSourceStage {
+
+  private implicit class ReceiveMessageExtension(request: ReceiveMessageRequest) {
```
I'd like to be it explicit when constructing the `ReceiveMessageRequest` instead.

barakbaronk commented:
@ennru thanks!
fixed your comment. I left the settings parameter as `Option` since zero finite duration is a valid value.
I also removed the implicit class, but I had to change the receive request variable to `var`

ennru commented:
Great that you reached out to improving Alpakka SQS - I hope your migration towards streaming integrations works well.
